week6/week6_prepartion_database/Program.cs

```diff
@@ -0,0 +1,271 @@
+(empty and the other) constructure 
+
+<example>
+ implement a simple endpoint that fetches the current users in MySql.
+  <questions>
+  what does the code do? So let’s analyze what we just did:
+  1.We created a new endpoint
+  2.Inside we created a new connection to MySql
+  3.We opened a connect
+  4.Then we created a new command with SQL text inside
+  5.We executed the command
+  6.In a loop we read rows and mapped columns to our model
+  7.Return the list to the caller
+
+  Then after:Change the code to select both host and user from the database:
+    1.Change the record for User and add string host
+    2.Change SQL to select host
+    3.Change the inner part of the loop to read user and host before creating a new object
+  </questions>
+</example>
+</summary>
+*/
+app.MapGet("/users", async (IConfiguration configuration) =>
+{
+  using (var connection = new MySqlConnection(configuration.GetConnectionString("Default")))
+  {
+    connection.Open();
+    using (var command = new MySqlCommand("SELECT * FROM mysql.user", connection))
+    {
+      using (var reader = await command.ExecuteReaderAsync())
+      {
+        var users = new List<User>();
+        while (await reader.ReadAsync())
+        {
+          var user = reader.GetString(reader.GetOrdinal("user"));
+          var host = reader.GetString(reader.GetOrdinal("host"));
+          users.Add(new User(user, host));
+        }
+        return Results.Ok(users);
+      }
+    }
+  }
+});
+
+/*
+< summary >
+Endpoint for getgetting users. starting from C# 8 it no longer need inner {}
+</summary>
+*/
+app.MapGet("/users2", async (IConfiguration configuration) =>
+{
+  // starting from C# 8 you no longer need inner {}
+  using var connection = new MySqlConnection(configuration.GetConnectionString("Default"));
+  var users = await connection.QueryAsync<User>("SELECT * FROM mysql.user");
+  return Results.Ok(users);
+});
+
+/*
+< summary >
+Endpoint for getgetting users
+</summary>
+*/
+app.MapGet("/products", async (IConfiguration configuration) =>
+{
+  using var connection = new MySqlConnection(configuration.GetConnectionString("Default"));
+  var products = await connection.QueryAsync<Product>("SELECT id, name, price,description FROM dapper.products");
+  return Results.Ok(products);
+});
+
+
+/*
+<question>
+    Implement the endpoint /product/{id}that:
+    1.Changes the SQL to add WHERE id = {id}
+    2.Instead of a list, returns .FirstOrDefault
+    3.Bonus: if there is no such user, the list would be empty. What should REST API return when you want a resource that doesn’t exist? E.g. you call it with /product/666?
+    4.Bonus: if the user passes negative number e.g. -1, what should REST API return in that case?
+</questions>
+*/
+app.MapGet("/products/{id}", async (IConfiguration configuration, int id) =>
+{
+  if (id < 0)
+  {
+    //#####
+    throw new Exception("id can not be negative");//??
+    // Results.BadRequest($" you entered id:{id},id can not be negative"); 
```

Comment:
I would prefer the BadRequest than Exception. What is the difference if you call it with -1?

Reply:
using badrequest  and calling with -id , it returns the Result.NotFound which is at the end of the at line 103. it does not print line 97 how to handle this then ?